xmtp_mls/src/storage/encrypted_store/group_message.rs

```diff
@@ -18,12 +25,282 @@ pub struct StoredGroupMessage {
     /// Time in nanoseconds the message was sent.
     pub sent_at_ns: i64,
     /// Group Message Kind Enum
-    pub kind: i32,
+    pub kind: GroupMessageKind,
     /// The ID of the App Installation this message was sent from.
     pub sender_installation_id: Vec<u8>,
     /// Network wallet address of the Sender
     pub sender_wallet_address: String,
 }
 
+#[repr(i32)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, AsExpression, FromSqlRow)]
+#[diesel(sql_type = Integer)]
+pub enum GroupMessageKind {
+    Application = 1,
+    MemberAdded = 2,
+    MemberRemoved = 3,
+}
+
+impl ToSql<Integer, Sqlite> for GroupMessageKind
+where
+    i32: ToSql<Integer, Sqlite>,
+{
+    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
+        out.set_value(*self as i32);
+        Ok(IsNull::No)
+    }
+}
+
+impl FromSql<Integer, Sqlite> for GroupMessageKind
+where
+    i32: FromSql<Integer, Sqlite>,
+{
+    fn from_sql(bytes: <Sqlite as Backend>::RawValue<'_>) -> deserialize::Result<Self> {
+        match i32::from_sql(bytes)? {
+            1 => Ok(GroupMessageKind::Application),
+            2 => Ok(GroupMessageKind::MemberAdded),
+            3 => Ok(GroupMessageKind::MemberRemoved),
+            x => Err(format!("Unrecognized variant {}", x).into()),
+        }
+    }
+}
+
 impl_fetch!(StoredGroupMessage, group_messages, Vec<u8>);
 impl_store!(StoredGroupMessage, group_messages);
+
+impl EncryptedMessageStore {
+    /// Query for group messages
+    pub fn get_group_messages<GroupId: AsRef<[u8]>>(
+        &self,
+        conn: &mut DbConnection,
+        group_id: GroupId,
+        sent_after: Option<i64>,
+        sent_before: Option<i64>,
+        kind: Option<GroupMessageKind>,
+    ) -> Result<Vec<StoredGroupMessage>, StorageError> {
+        use super::schema::group_messages::dsl;
+
+        let mut query = dsl::group_messages
+            .filter(dsl::group_id.eq(group_id.as_ref()))
+            .into_boxed();
+
+        if let Some(sent_after) = sent_after {
+            query = query.filter(dsl::sent_at_ns.gt(sent_after));
+        }
+
+        if let Some(sent_before) = sent_before {
+            query = query.filter(dsl::sent_at_ns.lt(sent_before));
+        }
+
+        if let Some(kind) = kind {
+            query = query.filter(dsl::kind.eq(kind));
+        }
+        Ok(query.load::<StoredGroupMessage>(conn)?)
+    }
+
+    /// Get a particular group message
+    pub fn get_group_message<MessageId: AsRef<[u8]>>(
```

Comment:
How is this different from the `Find` trait?

Reply:
First time I'm seeing `Find`, but it relates specifically to `diesel` ORM and the `find` method, tbh not deeply familiar with diesel ORM, but it seems at first glance the `FindDsl` trait would be most useful if you want to be generic over the key type. We could maybe experiment with refactoring the `Fetch` trait to make use of it somehow, and possibly remove the `Key` associated type. 

AsRef is mostly ergonomic/api -- it lets us pass a `&[u8]` or `Vec<u8>` without .to_vec() or constraining the type to something in that regard. It helps reduce cloning, and allows us to write

```rust
let id = vec![0x0];
// these are all valid
get_group_message(id)
get_group_message(id.as_slice());
get_group_message(&id);
```

It is a conversion trait, but meant to be v. cheap at-runtime

Reply:
Sorry sorry sorry, I meant the `Fetch` trait. Although this is pretty slick and I'm now sold on the value of keeping it as-is.